### PR TITLE
feat: Functional payload builders for add_rewards and transfer + full reporting on performAction calls

### PR DESCRIPTION
**Pull request by Tritium-VLK**

### CODE REVIEW

**Comment by Tritium-VLK** [comment]
Note I am a bit stuck here.  The action setup for merge_json and add_reward seems very similar/identical, yet somehow the environment variables are not making it through to the report runs on the add_reward.

@gosuto-inzasheru @jalbrekt85 any ideas?

[Example merge tthat works with env vars](https://github.com/BalancerMaxis/multisig-ops/actions/runs/9504455916)
[Example add_reward that fails to report due to missing env vars](https://github.com/BalancerMaxis/multisig-ops/actions/runs/9504869788)

**Comment by gosuto-inzasheru** [comment]
im not sure, but in the successful one you define the outputs of the first job explicitly:
```
    outputs:
      pull-request-number: ${{ steps.cpr.outputs.pull-request-number}}
      pull-request-head-sha: ${{ steps.cpr.outputs.pull-request-head-sha }}
```

in the unsuccessful one however, i can only find this:
```
    - name: Set outputs
      id: outputs
      run: |
        echo "::set-output name=pull-request-number::${{ steps.cpr.outputs.pull-request-number }}"
        echo "::set-output name=pull-request-branch::${{ steps.cpr.outputs.pull-request-branch }}"
```

not sure where that last bit comes from, but i think chatgpt ran out of coffee there 😋 

i would suggest to just define them explicitly in the same manner as in the merge_jsons job

**Comment by gosuto-inzasheru** [comment]
also `pull-request-branch` is just another halucination by chatgpt

[the outputs from that workflow are](https://github.com/peter-evans/create-pull-request?tab=readme-ov-file#action-outputs):
- `pull-request-number`
- `pull-request-url`
- `pull-request-operation`
- `pull-request-head-sha`

**Comment by Tritium-VLK** [comment]
> also `pull-request-branch` is just another hallucination by chatgpt
> 
> [the outputs from that workflow are](https://github.com/peter-evans/create-pull-request?tab=readme-ov-file#action-outputs):
> 
> * `pull-request-number`
> * `pull-request-url`
> * `pull-request-operation`
> * `pull-request-head-sha`

I was hallucinating with her.  The issue here is that we don't want to/can't use a static branch name because there may be more than one open at a time.  So we need some way to pass a ref that can be checked out and commited to on to the report action.  will try the head-sha.

> also `pull-request-branch` is just another hallucination by chatgpt
> 
> [the outputs from that workflow are](https://github.com/peter-evans/create-pull-request?tab=readme-ov-file#action-outputs):
> 
> * `pull-request-number`
> * `pull-request-url`
> * `pull-request-operation`
> * `pull-request-head-sha`

**Comment by Tritium-VLK** [comment]
add_reward example: #1065 
transfer example: #1066

**Comment by Tritium-VLK** [comment]
Note that I also moved all of the prettification stuff to one place (or what I found) and tried to use it as much as possible.  This means we can define formatting there.  The linefeeds/stringification of list members I am still doing outside the prettification functions.  Maybe good to think about where that makes sense.

File: .github/workflows/create_add_rewards_payload.yaml [hunk 1]
```diff
@@ -32,32 +34,55 @@ jobs:
...
+          GAUGE: ${{ github.event.inputs.gauge }}
+          DISTRIBUTOR: ${{ github.event.inputs.distributor }}
+          MULTISIG: ${{ github.event.inputs.multisig }}
+          CHAIN_NAME: ${{ github.event.inputs.chain_name }}
+
+
+      - name: Generate branch name
+        id: generate-branch-name
+        run: |
+          echo "::set-output name=branch::${{ github.job }}-$(date +'%Y%m%d%H%M%S')"
```
Comment (gosuto-inzasheru) [inline .github/workflows/create_add_rewards_payload.yaml:70]: what is the reason for using this way of creating the output, vs `outputs:`?

**Review by gosuto-inzasheru — COMMENTED** [review]
so if i understand correctly this creates a payload that does auth.performAction(add_rewards(_reward_token, _distributor))`. then there is also a custom report that reads auth payloads and can report on it, but is limited to `add_rewards()` only

doing such actions through tooling (if necessary) is handy, but how does it relate to already existing tools such as https://defilytica.tools/#/balancer/payloadBuilder? what is a usecase you envision for this?

i think a deeper discussion around tooling and for example gh actions vs react vs python is needed before starting to build this. because if there was such a discussion i missed it.

File: action-scripts/brownie/scripts/report_gauges.py [hunk 2]
```diff
@@ -524,6 +528,73 @@ def _parse_permissions(transaction: dict, **kwargs) -> Optional[dict]:
...
+    except Exception as e:
+        print(f"Error processing performAction call: {e}")
+        return
+    try:
+        inputs = prettify_flat_list(inputs, chain_name)
+    except Exception as e:
+        print(
+            f"INFO: Was unable to prettify inputs during performAction decoding, perhaps the input data is too complex: {e}"
+        )
+    return {
```
[the remark this answers is not in the document]
Comment (Tritium-VLK) [reply action-scripts/brownie/scripts/report_gauges.py:587]: It was written to use from_explorer and try to work with any call_data.  Gauges somehow have issues, based on how they are factory deployed, so I just put in a catch for that.  It should work in many cases already, and be easy to expand for more exceptions with some kind of a try/catch/probe jobby similar to what is used in the payload report.

File: .github/workflows/create_add_rewards_payload.yaml (hunk 1 above)
Comment (Tritium-VLK) [reply .github/workflows/create_add_rewards_payload.yaml:70]: I spent a bunch of time fighting with it, some vars you can't pass into the PR action, in that context.  This ended up working.  If you think you can figure out another way to get the job name and the date in the branch: section be my guest.  This step is just using bash to create an output I can read in in Create PR.

**Comment by Tritium-VLK** [comment]
Merging main back in was a bit complex, this has become a wide delta.  Twiddling a bunch more payloads for a final test to make sure nothing is broken.

**Comment by Tritium-VLK** [comment]
Rollback test after some failures in other stuff.  Also need to add target to report.

**Comment by Tritium-VLK** [comment]
@gosuto-inzasheru this would be useful for this weeks governance.  I kind of think we should just go ahead and merghe it and deal with whatever breaks/ if it breaks otherwise.  If you agree pls approve and lets go :)

**Review by gosuto-inzasheru — APPROVED** [review]
decoding of add_rewards payload is valueable and worth the merge

not sure of the value of payload building workflows but that is a separate discussion

**Pull request merged**